src/common/wallet-state.test.ts

```diff
@@ -38,25 +39,67 @@ const pending1 = createTestTx('outgoing', 'pending', 100, 10)
 const pending2 = createTestTx('outgoing', 'pending', 200, 10)
 const incomingPending = createTestTx('incoming', 'pending', 300, 10)
 
-it('calculates pending balance correctly in case of no pending txns', () => {
-  assert.deepEqual(getPendingBalance(pendingTestBaseTransactions), asEther(0))
-})
+describe('pending balance calculation', () => {
+  it('calculates pending balance correctly in case of no pending txns', () => {
+    assert.deepEqual(getPendingBalance(pendingTestBaseTransactions), asEther(0))
+  })
 
-it('calculates pending balance correctly in case of 1 pending txn', () => {
-  assert.deepEqual(getPendingBalance([...pendingTestBaseTransactions, pending1]), asEther(110))
-})
+  it('calculates pending balance correctly in case of 1 pending txn', () => {
+    assert.deepEqual(getPendingBalance([...pendingTestBaseTransactions, pending1]), asEther(110))
+  })
 
-it('calculates pending balance correctly in case of multiple pending txns', () => {
-  assert.deepEqual(
-    getPendingBalance([...pendingTestBaseTransactions, pending1, pending2]),
-    asEther(320),
-  )
+  it('calculates pending balance correctly in case of multiple pending txns', () => {
+    assert.deepEqual(
+      getPendingBalance([...pendingTestBaseTransactions, pending1, pending2]),
+      asEther(320),
+    )
+  })
+
+  it('calculates pending balance correctly in case of an incoming pending txn', () => {
+    // Adding an incoming pending tx doesn't make a difference
+    assert.deepEqual(
+      getPendingBalance([...pendingTestBaseTransactions, pending1, pending2, incomingPending]),
+      asEther(320),
+    )
+  })
 })
 
-it('calculates pending balance correctly in case of an incoming pending txn', () => {
-  // Adding an incoming pending tx doesn't make a difference
-  assert.deepEqual(
-    getPendingBalance([...pendingTestBaseTransactions, pending1, pending2, incomingPending]),
-    asEther(320),
-  )
+describe('transaction history merging', () => {
+  it('marks disappearing pending transactions as failed', () => {
+    const previous = [...pendingTestBaseTransactions, pending1]
+    const current = pendingTestBaseTransactions
+
+    const result = mergeTransactions(current)(previous)
+
+    assert.deepEqual(result, [{...pending1, status: 'failed'}, ...pendingTestBaseTransactions])
+  })
+
+  it('keeps failed transactions', () => {
+    const failed: Transaction = {...pending1, status: 'failed'}
+    const previous = [...pendingTestBaseTransactions, failed]
+    const current = pendingTestBaseTransactions
+
+    const result = mergeTransactions(current)(previous)
+
+    assert.deepEqual(result, [failed, ...pendingTestBaseTransactions])
+  })
+
+  it('keeps disappearing confirmed and persisted transactions', () => {
```

Comment:
Shouldn't this be actually 'keeps confirmed and persisted transactions'? Actually, on the fourth read, I understand what it tries to say, but can be understood both ways - (keeps disappearing) (confirmed and persisted transactions). Maybe we can go with 'keeps confirmed and persisted transactions which disappeared'?